test/network/network_test.cpp

```diff
@@ -0,0 +1,366 @@
+#pragma GCC diagnostic ignored "-Wconversion"
+
+#include <sys/socket.h>
+#include <sys/types.h>
+#include <util/test_harness.h>
+#include <pqxx/pqxx> /* libpqxx is used to instantiate C++ client */
+
+#include <cstdio>
+#include <cstring>
+#include <string>
+#include <vector>
+#include "common/settings.h"
+#include "gtest/gtest.h"
+#include "loggers/main_logger.h"
+#include "network/connection_handle_factory.h"
+
+#define NUM_THREADS 1
+
+/*
+ * Read and write buffer size for the test
+ */
+#define TEST_BUF_SIZE 1000
+
+namespace terrier::network {
+
+//===--------------------------------------------------------------------===//
+// Simple Query Tests
+//===--------------------------------------------------------------------===//
+
+class NetworkTests : public TerrierTest {
+ protected:
+  TerrierServer server;
+  uint16_t port = common::Settings::SERVER_PORT;
+  std::thread server_thread;
+
+  /**
+   * Initialization
+   */
+  void SetUp() override {
+    TerrierTest::SetUp();
+
+    network_logger->set_level(spdlog::level::debug);
+    spdlog::flush_every(std::chrono::seconds(1));
+
+    try {
+      server.SetPort(port);
+      server.SetupServer();
+    } catch (NetworkProcessException &exception) {
+      LOG_ERROR("[LaunchServer] exception when launching server");
+      throw;
+    }
+    LOG_DEBUG("Server initialized");
+    server_thread = std::thread([&]() { server.ServerLoop(); });
+  }
+
+  void TearDown() override {
+    server.Close();
+    server_thread.join();
+    LOG_DEBUG("Terrier has shut down");
+
+    TerrierTest::TearDown();
+  }
+};
+
+/**
+ * Use std::thread to initiate peloton server and pqxx client in separate
+ * threads
+ * Simple query test to guarantee both sides run correctly
+ * Callback method to close server after client finishes
+ */
+// NOLINTNEXTLINE
+TEST_F(NetworkTests, SimpleQueryTest) {
+  try {
+    pqxx::connection C(
+        fmt::format("host=127.0.0.1 port={0} user=postgres sslmode=disable application_name=psql", port));
+
+    pqxx::work txn1(C);
+    txn1.exec("INSERT INTO employee VALUES (1, 'Han LI');");
+    txn1.exec("INSERT INTO employee VALUES (2, 'Shaokun ZOU');");
+    txn1.exec("INSERT INTO employee VALUES (3, 'Yilei CHU');");
+
+    pqxx::result R = txn1.exec("SELECT name FROM employee where id=1;");
+    txn1.commit();
+    EXPECT_EQ(R.size(), 0);
+  } catch (const std::exception &e) {
+    LOG_ERROR("[SimpleQueryTest] Exception occurred: {0}", e.what());
+    EXPECT_TRUE(false);
+  }
+  LOG_DEBUG("[SimpleQueryTest] Client has closed");
+}
+
+ssize_t ReadUntilReadyOrClose(char *in_buffer, size_t max_len, int socket_fd) {
+  ssize_t n;
+  while (true) {
+    n = read(socket_fd, in_buffer, max_len);
+    if (n == 0 || in_buffer[n - 6] == 'Z')  // Ready for request
```

Comment:
If n > 0 but n < 6, this is going to produce an invalid array index.

Reply:
@DarkForte Could you look at this